tools/bbmap.py

```diff
@@ -28,38 +28,94 @@ def __init__(self, install_methods=None):
             tmp_bam = os.path.join(t_dir, 'bbmap_out.bam')
-            self.execute(tool='bbmap.sh', in1=in1, in2=in2, ref=refFasta, out=tmp_bam, nodisk=nodisk, **kwargs)
+            self.execute(tool='bbmap.sh', in1=in1, in2=in2, ref=ref_fasta, out=tmp_bam, nodisk=nodisk, JVMmemory=JVMmemory, **kwargs)
             
             # Samtools filter (optional)
             if min_qual:
+                samtools = tools.samtools.SamtoolsTool()
                 tmp_bam2 = os.path.join(tdir, 'bbmap.filtered.bam')
-                cmd = [samtools.install_and_get_path(), 'view', '-b', '-S', '-1', '-q', str(min_qual), tmp_bam]
-                _log.debug('%s > %s', ' '.join(cmd), tmp_bam2)
-                with open(tmp_bam2, 'wb') as outf:
-                    util.misc.run_and_save(cmd, outf=outf)
+                samtools.view(['-b', '-S', '-1', '-q', str(min_qual)], tmp_bam, tmp_bam2)
                 os.unlink(tmp_bam)
                 tmp_bam = tmp_bam2
 
             # Picard SortSam
             sorter = tools.picard.SortSamTool()
             sorter.execute(
                 tmp_bam,
-                outBam,
+                out_bam,
                 sort_order='coordinate',
                 picardOptions=['CREATE_INDEX=true', 'VALIDATION_STRINGENCY=SILENT'],
                 JVMmemory=JVMmemory
             )
 
+    def dedup_clumpify(self, in_bam, out_bam, optical=False, subs=5, passes=4, dupedist=40, kmer_size=31, spany=False, adjacent=False, treat_as_unpaired=False, JVMmemory=None, **kwargs):
+        '''
+            clumpify-based deduplication
+            see:
+                https://www.biostars.org/p/225338/
+                https://www.biostars.org/p/225338/#230178
+            and also:
+                https://www.biostars.org/p/229842/#229940
+                https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/clumpify-guide/
+
+            From clumpify.sh usage:
+                optical=False   If true, *only* mark or remove *optical* duplicates.
+                dedupe=True     Remove duplicate reads.  For pairs, both must match.
+                dupedist=40     (dist) Max distance to consider for optical duplicates.
+                                Higher removes more duplicates but is more likely to
+                                remove PCR rather than optical duplicates.
+                                This is platform-specific; recommendations:
+                                   NextSeq      40  (and spany=t)
+                                   HiSeq 1T     40
+                                   HiSeq 2500   40
+                                   HiSeq 3k/4k  2500
+                                   Novaseq      12000
+                k=31            Use kmers of this length (1-31).  Shorter kmers may
+                                increase compression, but 31 is recommended for error
+                                correction.                
+    '''
```

Comment:
indent

Reply:
Possibly add comment that optical duplicates is weaker dedupe limited by xy-position. Meant for single-end sequencing like that commonly used in NextSeq